### ISSUE TITLE
Viewing Apollo images in Rviz

### ISSUE THREAD

**Issue by snuffysasa** [ApolloAuto/apollo#4552]
Hello, 

I am new to Apollo. And I am wondering is there away to view the Apollo camera images in Rviz?

with the image topic? 
I am referring to images like this:
/apollo/sensor/camera/obstacle/front_6mm

In Rviz I get an error like:  yuyv unsupported image encoding

Is there anyway to deal with this?

Thank you

**Comment by jilinzhou** [comment]
Apollo uses google protobuf for its message exchanges which are different from standard ROS msg types. Basically you need recompile RVIZ agaist apollo ros stack to make it work. Or you can recast apollo camera message types to standard ros camera msg type.

**Comment by snuffysasa** [comment]
@jilinzhou 

"Apollo uses google protobuf for its message exchanges which are different from standard ROS msg types."

Thank you,

So is this true for all Apollo messages not just the image messages?

"Or you can recast apollo camera message types to standard ros camera msg type."

Can I build my own ROS node that subscribes to one or many Apollo messages, then republishes them as standard ROS messages?    Is there any information, or guides, or examples of this?  I am assuming there is not a guide, but is there anything I can reference somewhere in Apollo that would be helpful in doing this?

Thank you much

**Comment by jilinzhou** [comment]
@snuffysasa 

So is this true for all Apollo messages not just the image messages?

Yes. 

Can I build my own ROS node that subscribes to one or many Apollo messages, then republishes them as standard ROS messages? Is there any information, or guides, or examples of this? I am assuming there is not a guide, but is there anything I can reference somewhere in Apollo that would be helpful in doing this?

Yes. You even can directly modify interested modules so they publish standard ROS msgs directly to avoid some traffic overhead. To do it:

1. Locate the source code which create your interested protobuf msgs.
2. Replace with the code to create standard ros msg types and publish.

You many need to modify AdapterManager (not sure) to publish the standard ros mgs. If you do need, the work should be trivial.

**Comment by techoe** [comment]
Thank you @jilinzhou !

**Comment by yonhdee** [comment]
you can actually subscribe this topic outside docker, and write your own topic recaster node for rviz to visualizer. check the original usb_cam package, there're some lines doing the yuyv2bgr format convertion, won't take too long

**Comment by snuffysasa** [comment]
@jilinzhou 

Hello, Sorry to dig this up. I am revisiting this issue.

I have done the steps you mention. "You even can directly modify interested modules so they publish standard ROS msgs directly to avoid some traffic overhead. "  I did this ant it works, thank you.  

But now, I would like to create my own custom Ros msg that is not a proto msg.   And now I am stumped.. because I can not figure out how to compile my own .msg with Bazel Build and have it generate the header files, so that I can include them.  Any ideas on how to make my own custom msg that is not a proto msg in the apollo docker project.


@yonhdee 

thank you,

I noticed this too.  So is the /apollo/sensor/camera/obstacle/front_6mm topic... not a protobuf msg, and just a standard ros sensor Image msg, that is in yuv format?   Or am I misunderstanding
